cpp/src/copying/get_element.cu

```diff
@@ -14,128 +14,114 @@
-  std::unique_ptr<scalar>> operator()(
-    column_view const& input,
+  template <typename T>
+  std::enable_if_t<is_fixed_width<T>(), std::unique_ptr<scalar>>
+  operator()(
+    column_view const &input,
     size_type index,
-    cudaStream_t stream = 0,
-    rmm::mr::device_memory_resource *mr = rmm::mr::get_default_resource())
-  {
+    cudaStream_t stream                 = 0,
+    rmm::mr::device_memory_resource *mr = rmm::mr::get_default_resource()) {
     auto s = make_fixed_width_scalar(data_type(type_to_id<T>()), stream, mr);
 
     using ScalarType = cudf::experimental::scalar_type_t<T>;
-    auto typed_s = static_cast<ScalarType *>(s.get());
-    
-    auto device_s = get_scalar_device_view(*typed_s);
+    auto typed_s     = static_cast<ScalarType *>(s.get());
+
+    auto device_s   = get_scalar_device_view(*typed_s);
     auto device_col = column_device_view::create(input, stream);
-    
-    single_thread_kernel<<<1, 1, 0, stream>>>(
-      [ device_s, d_col=*device_col, index ] __device__ () mutable {
+
+    device_single_thread(
+      [device_s, d_col = *device_col, index] __device__() mutable {
         device_s.set_value(d_col.element<T>(index));
         device_s.set_valid(d_col.is_valid(index));
-      });
+      },
+      stream);
     return s;
   }
 
-  template<typename T>
-  std::enable_if_t<std::is_same<T, string_view>::value, 
-  std::unique_ptr<scalar>> operator()(
-    column_view const& input,
+  template <typename T>
+  std::enable_if_t<std::is_same<T, string_view>::value, std::unique_ptr<scalar>>
+  operator()(
+    column_view const &input,
     size_type index,
-    cudaStream_t stream = 0,
-    rmm::mr::device_memory_resource *mr = rmm::mr::get_default_resource())
-  {
+    cudaStream_t stream                 = 0,
+    rmm::mr::device_memory_resource *mr = rmm::mr::get_default_resource()) {
     auto device_col = column_device_view::create(input, stream);
 
     rmm::device_scalar<string_view> temp_data;
     rmm::device_scalar<bool> temp_valid;
 
-    single_thread_kernel<<<1, 1, 0, stream>>>(
-      [ 
-        buffer=temp_data.data(),
-        validity=temp_valid.data(),
-        d_col=*device_col,
-        index 
-      ] __device__ () mutable {
-        *buffer = d_col.element<string_view>(index);
+    device_single_thread(
+      [buffer   = temp_data.data(),
+       validity = temp_valid.data(),
+       d_col    = *device_col,
+       index] __device__() mutable {
+        *buffer   = d_col.element<string_view>(index);
         *validity = d_col.is_valid(index);
-      });
+      },
+      stream);
 
-    return std::make_unique<string_scalar>(
-      temp_data, temp_valid.value(stream), stream, mr);
+    return std::make_unique<string_scalar>(temp_data, temp_valid.value(stream), stream, mr);
   }
 
-  template<typename T>
-  std::enable_if_t<std::is_same<T, dictionary32>::value, 
-  std::unique_ptr<scalar>> operator()(
-    column_view const& input,
+  template <typename T>
```

Comment:
TBH I prefer SFINAE in the template rather than the return type. The latter messes up doxygen when the return type definition gets very long (since doxygen won't wrap it).  But we have a lot of the former in libcudf already so up to you.

```suggestion
template <typename T, std::enable_if_t<not cudf::is_same<T, dictionary32>::value>* = nullptr>
std::unique_ptr<scalar> operator()(...
```

Reply:
I get the following error, if I used SFINAE on template.
```
cpp/src/copying/get_element.cu: In member function ‘std::unique_ptr<cudf::scalar> cudf::experimental::detail::_GLOBAL__N__46_tmpxft_000008e9_00000000_6_get_element_cpp1_ii_83c1900b::get_element_functor::operator()(const cudf::column_view&, cudf::size_type, cudaStream_t, rmm::mr::device_memory_resource*)’:
cpp/src/copying/get_element.cu:97:246: error: ‘__T292’ was not declared in this scope
cpp/src/copying/get_element.cu: In instantiation of ‘std::unique_ptr<cudf::scalar> cudf::experimental::detail::_GLOBAL__N__46_tmpxft_000008e9_00000000_6_get_element_cpp1_ii_83c1900b::get_element_functor::operator()(const cudf::column_view&, cudf::size_type, cudaStream_t, rmm::mr::device_memory_resource*) [with T = cudf::dictionary_wrapper<int>; std::enable_if_t<std::is_same<T, cudf::dictionary_wrapper<int> >::value, void>* <anonymous> = 0u; cudf::size_type = int; cudaStream_t = CUstream_st*]’:
cpp/include/cudf/utilities/type_dispatcher.hpp:317:103:   required from ‘constexpr decltype(auto) cudf::experimental::type_dispatcher(cudf::data_type, Functor, Ts&& ...) [with IdTypeMap = cudf::experimental::id_to_type_impl; Functor = cudf::experimental::detail::_GLOBAL__N__46_tmpxft_000008e9_00000000_6_get_element_cpp1_ii_83c1900b::get_element_functor; Ts = {const cudf::column_view&, int&, CUstream_st*&, rmm::mr::device_memory_resource*&}]’
cpp/src/copying/get_element.cu:114:93:   required from here
cpp/src/copying/get_element.cu:97:21: error: could not convert template argument ‘&((cudf::experimental::detail::_GLOBAL__N__46_tmpxft_000008e9_00000000_6_get_element_cpp1_ii_83c1900b::get_element_functor*)this)->*operator()<<template arguments error> >’ to ‘std::unique_ptr<cudf::scalar> (cudf::experimental::detail::_GLOBAL__N__46_tmpxft_000008e9_00000000_6_get_element_cpp1_ii_83c1900b::get_element_functor::*)(const cudf::column_view&, int, CUstream_st*, rmm::mr::device_memory_resource*’
     device_single_thread(
                     ^
CMakeFiles/cudf.dir/build.make:2341: recipe for target 'CMakeFiles/cudf.dir/src/copying/get_element.cu.o' failed
```

Reply:
This is the nvcc device lambda restriction in action.

Reply:
Issue solved by giving a name to the second template param.
```suggestion
  template <typename T, std::enable_if_t<cudf::is_same<T, dictionary32>::value>p* = nullptr>
  std::unique_ptr<scalar> operator()(
```